Documentation, access to the fuzz targets inside serenity, crash in file manager

BenWiederhake opened the pull request:
- I had to set up serenity from scratch and ironed out a few shortcomings in the QtCreator setup documentation
- I *again* ran into the issue of wanting to "just" run the FuzzBMPLoader code inside Serenity. There used to be no way to do this, now there is.
- The bug wasn't in the BMP loader at all, instead it was in LibGUI: When making a thumbnail, don't use the exact same format unconditionally, especially since the later stages assume (A)RGB32.

alimpfard commented on Userland/Utilities/test-fuzz.cpp:
```diff
@@ -0,0 +1,130 @@
+ENUMERATE_TARGETS(__ENUMERATE_TARGET)
+#undef __ENUMERATE_TARGET
+
+// FIXME: How to automate this?
```
Generate this particular file, maybe?

BenWiederhake replied:
Hmm, that would involve non-trivial amounts of cmake magic. Also, we might exclude some fuzzers, for example FuzziliJS, so it shouldn't be anything that tries to take *all* fuzzers.

Ironically, `#include` *does* work with macro-built filenames, but the `#define` doesn't. Just in case that sparks an idea.

alimpfard replied:
Why build it with cmake? we have a bunch of generators written in C++ already.

Hmm...sure, switch to this (not sure if `#include` works like this, but worth a try)
```
#define INCLUDE_FUZZER(name) \
    #define LLVMTestOneInput run
    namespace name { \
    #include wherever/name/is \
    }
```

BenWiederhake replied:
This will fail for three reasons:

1. A macro can't include a `#`, because it will be interpreted as "stringify the macro argument with the following name", and then complain about `include` not being a macro argument. That's why I keep saying that a macro can't `#include`. Do note that a `#include` can make use of macros.
2. `namespace` will cause lots of trouble, because the fuzz targets in turn `include` a bunch of headers, which cause the fuzz target try to link against a symbol `name::Gfx::BMPLoader::…`, which doesn't exist.
3. `namespace` doesn't even solve the problem, because the fuzzers are (and have to be) `extern "C"`, so the namespacing is effectively ignored.

alimpfard replied:
oh yikes. 
It seems I misunderstood your "#include does work with macro-built filenames", my bad.

Then I think it's either listing them all, or accepting the downsides and generating them (at least partially, maybe just `#define INCLUDED_AS_UTILITY` and just put all of FuzziliJs behind an `#ifndef INCLUDED_AS_UTILITY`)

BenWiederhake replied:
That doesn't work as easily either, because the decision made in FuzziliJS would somehow need to travel to the output of `test-fuzz list all`, or else that command would list FuzziliJS as a possible target.

alimpfard replied:
Fair enough. so it's either a generator that can make the decisions or just listing them all.
these aren't that frequently updated, so maybe it being manual isn't _that_ bad...

alimpfard commented on Userland/Utilities/test-fuzz.cpp:
```diff
@@ -0,0 +1,130 @@
+//    T(Shell) \
+//    T(TTF) \
+//    T(URL)
+*/
```
Any particular reason why these are disabled?

BenWiederhake replied:
Nope, just laziness.

BenWiederhake replied:
Converted to draft until I finish that part.

alimpfard replied:
No real reason to block on that, just wondering if there was some issue with those specific fuzzers.

alimpfard commented on Userland/Utilities/test-fuzz.cpp:
```diff
@@ -0,0 +1,130 @@
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(type, "Which type of fuzzer you'd like to try (use \"list\" to list all existing)", "fuzz-kind");
+    args_parser.add_positional_argument(filename, "Which file to use as input", "filename");
```
I think it's more common to write something like "Type of fuzzer to run" and "Fuzzer input file" instead?
Just a nit

alimpfard commented on Userland/Utilities/test-fuzz.cpp:
```diff
@@ -0,0 +1,130 @@
+
+typedef int (*TestOneInputFn)(const uint8_t*, size_t);
+
+static TestOneInputFn parse_target_name(const String& name)
```
`auto` should be able to infer this, I think.
Below too, so you can very likely omit the typedef above

BenWiederhake replied:
I didn't know that `auto` works for return types!

alimpfard commented on Userland/Utilities/test-fuzz.cpp:
```diff
@@ -0,0 +1,130 @@
+    args_parser.add_positional_argument(filename, "Which file to use as input", "filename");
+    args_parser.parse(argc, argv);
+
+    TestOneInputFn fn = parse_target_name(type);
```
(`auto` here)

BenWiederhake commented:
Changes since first push:
- Many nits from @alimpfard – thanks!
- Extended it to *all* existing fuzzing targets (except FuzziliJS which works very differently)

I'm still uncertain about the many `#include`s. Compiling them as separate unit asks for trouble (how to resolve the obvious naming conflict?), and handling the looping by cmake code generation spreads out the definition even further – after all, the enumeration and the actually included fileset needs to be kept in sync.I can't think of anything better than manually listing all the included files, so that's what I'll go with. Suggestions welcome.

awesomekling commented:
Is that file manager bug the same as #5299 perhaps?

BenWiederhake commented:
Technically yes, because now it crashes somewhere else.

BenWiederhake commented:
Changes since the last push:
- Sneakily remove a "FIXME" in `Userland/Utilities/test-fuzz.cpp`, because apparently there's no way to do this any better.
- Rebase for convenience.

awesomekling commented:
> Technically yes, because now it crashes somewhere else.

Does that make "Undocumented bug" true or false?

BenWiederhake commented:
Right, I should change that, too ^^'

Changes since last push:
- Reword that commit message
- Rebase
